### ISSUE TITLE
Integrate Next Gen Keto with wider policy ecosystems / Open Policy Agent

### ISSUE THREAD

**Issue by robinbraemer** [ory/keto#318]
Maybe it is already obvious to some, or maybe I can present some new viewpoints, but I wanted to write it down and formalize my thought experiment to everyone.

> For those who want to know more about Open Policy Agent, I recommend this video: https://youtu.be/Vdy26oA3py8

Looking ahead, as part of Next Gen Keto's efforts to empower wider ecosystems to configure policy and access control, specifically Access Control Lists and other authorization models, in a more scalable and distributed way, Next Gen Keto will not only be able to be used as a single service for access control and be the base authorization system other services depend and built on, but also **should** support integration with Open Policy Agent using OPA's extension functionality in that Keto's ACL system can be used to setup more various policies with OPA while leveraging Keto's powerful ACL system to fetch checks efficiently.

Keto ACL is not designed to be a complex policy engine, but allows efficient ACL checks at scale, that OPA cannot do (as built-in). ACL check requests in Next Gen Keto are still limited to it's small and simple set of parameters in which more complex policies cannot be defined like in OPA.

This makes the comparison of Next Gen Keto vs. OPA an oranges and apples comparison. Both systems are better in solving different aspects in the world of access control.

Businesses need a platform to define complex policies, extend and unit test 
their policy, which OPA is perfect for!

Adding Next Gen Keto into construct it can, as well, be put as an additional tool to [extend Open Policy Agent](https://www.openpolicyagent.org/docs/latest/extensions/)  functionality for providing scalable "object,relation,subject" ACLs.

This effectively means, OPA is like the platform that Keto can integrate with too. Much like Kubernetes is the operating system for Istio, while Istio could also be setup in a non-k8s environment.

Therefore we should provide OPA rego functions that users can use in OPA to perform checks via their Keto deployment as part of their policy setup.

**Comment by aeneasr** [comment]
Yes, this makes absolute sense. We have good a relationship with the OPA maintainers and I am sure that they would appreciate and potentially even support such an effort!

**Comment by zepatrik** [comment]
Nice, this actually goes hand in hand with my considerations in #319

**Comment by aeneasr** [comment]
It also makes sense because we could simply publish a tutorial for running OPA with the current Ory Keto policies, which would allow people to do a "soft" migration of current Ory Keto to another system that still works as they're used to.

**Comment by zepatrik** [comment]
Yes :heart_eyes: I had headaches because of migration paths for current users already.

**Comment by robinbraemer** [comment]
> Note taking comment:

Topaz is basically doing what I described in this issue, provide Rego functions to call a Zanzibar system as part of evaluation step:
https://www.topaz.sh/docs/intro#zanzibar--rebac

> Authorization policies can combine ABAC-style rules with ReBAC-style graph queries via a set of Rego built-ins.

**Comment by jeffsec-aws** [comment]
As Keto defines itself as:
> a policy decision point. It uses a set of access control policies, **similar to AWS IAM Policies,** in order to determine whether a subject (user, application, service, car, ...) is authorized to perform a certain action on a resource.

It would be interesting that, if the ecosystem opens to other Policy based Access Control, to consider [Cedar ](https://github.com/cedar-policy) as an extension to it as its opensource birth came for year of maintaining, operating, and evolving AWS IAM.

**Comment by mfreeman451** [comment]
Did the project ever decide on dealing with this inside Keto (OPA integration), or is this left as an excersize to the reader still? We are going to be using both in a project, just wondering if I can get them to talk to together or what.
